tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/beam_pipeline.py

```diff
@@ -0,0 +1,246 @@
+                 ----------------
+"""
+
+import copy
+import tensorflow as tf
+import apache_beam as beam
+
+
+@beam.ptransform_fn
+def ExecuteOneGraph(pcoll,  # pylint: disable=invalid-name
+                    op_to_execution_specs,
+                    op_to_remote_op_name_mapping,
+                    graph_name):
+  """Compute one graph, for example remote_op_b's graph.
+
+  Main assumption:
+    The parent graph has set up the placeholder inputs for the child graph.
+    (So we need to setup PColl for the main graph)
+
+  Args:
+    pcoll: input PCollection, each unit contains
+           {graph_name: {computed node: value}}
+    op_to_execution_specs: {graph_name: [spec1, spec2, ...]}
+    spec: {'subgraph': a graph_def,
+           'inputs': a set of node names,
+           'outputs': a set of node names,
+           'is_remote_op': a boolean}
+
+    # This information was stored in py_func's feed_dict,
+    # so we don't have it.
+    op_to_remote_op_name_mapping:
+      {graph_name:
+        {remote op name: {placeholder name inside subgraph: input name}}}
+
+    graph_name: which graph am I currently executing
+
+  Returns:
+    pcoll with the intermediate/end results of this graph.
+  """
+  execution_specs = _get_execution_specs(graph_name, op_to_execution_specs)
+
+  count = 0
+  for spec in execution_specs:
+    if not spec['is_remote_op']:
+      count += 1
+      pcoll = pcoll | str(count) >> beam.ParDo(_ExecuteOneSubgraph(),
+                                               spec,
+                                               graph_name)
+    else:
+      current_graph_name = graph_name
+      # remote op is the only output
+      remote_graph_name = list(spec['outputs'])[0]
+
+      count += 1
+      pcoll = pcoll | str(count) >> _LoadRemoteGraphInputs(
+          current_graph_name,
+          remote_graph_name,
+          op_to_remote_op_name_mapping)
+
+      # A good place to add beam.Reshuffle()
+      count += 1
+      pcoll = pcoll | str(count) >> ExecuteOneGraph(
+          op_to_execution_specs,
+          op_to_remote_op_name_mapping,
+          remote_graph_name)
+
+      count += 1
+      pcoll = pcoll | str(count) >> _ExtractRemoteGraphOutput(
+          current_graph_name,
+          remote_graph_name,
+          op_to_execution_specs)
+
+  return pcoll
+
+
+def _get_execution_specs(graph_name, op_to_execution_specs):
+  op_type = _get_op_type(graph_name, set(op_to_execution_specs.keys()))
+  return op_to_execution_specs[op_type]
+
+
+def _get_op_type(graph_name, ops):
+  """As an example, graph_name='remote_op_b_1' has op_type='remote_op_b'."""
+  for op_type in ops:
+    if graph_name.startswith(op_type):
+      return op_type
+  return ''
+
+
+class _ExecuteOneSubgraph(beam.DoFn):
+  """Compute one subgraph."""
+
+  def process(self, element, spec, graph_name):
+    """Executes the smallest unit: one subgraph.
+
+    Args:
+      element: a unit of PCollection, {graph_name: {computed node: value}}
```

Comment:
nit: delete "a unit of PCollection"

As Goutham mentioned earlier, consider making this a dataclass instead.

Reply:
Changed.